ci/docker-publish.yml

```diff
@@ -49,32 +56,31 @@ jobs:
       packages: write
 
     steps:
-      - uses: actions/checkout@v2
-
-      - name: Build image
-        run: docker build . --file Dockerfile --tag $IMAGE_NAME
-
-      - name: Log into registry
-        run: echo "${{ secrets.GITHUB_TOKEN }}" | docker login docker.pkg.github.com -u ${{ github.actor }} --password-stdin
-
-      - name: Push image
-        run: |
-          IMAGE_ID=docker.pkg.github.com/${{ github.repository }}/$IMAGE_NAME
-
-          # Change all uppercase to lowercase
-          IMAGE_ID=$(echo $IMAGE_ID | tr '[A-Z]' '[a-z]')
-
-          # Strip git ref prefix from version
-          VERSION=$(echo "${{ github.ref }}" | sed -e 's,.*/\(.*\),\1,')
-
-          # Strip "v" prefix from tag name
-          [[ "${{ github.ref }}" == "refs/tags/"* ]] && VERSION=$(echo $VERSION | sed -e 's/^v//')
-
-          # Use Docker `latest` tag convention
-          [ "$VERSION" == "$default-branch" ] && VERSION=latest
-
-          echo IMAGE_ID=$IMAGE_ID
-          echo VERSION=$VERSION
-
-          docker tag $IMAGE_NAME $IMAGE_ID:$VERSION
-          docker push $IMAGE_ID:$VERSION
+      - name: Checkout repository
+        uses: actions/checkout@v2
+
+      - name: Log into registry ${{ env.REGISTRY }}
+        uses: docker/login-action@28218f9b04b4f3f62068d7b6ce6ca5b26e35336c
+        with:
+          registry: ${{ env.REGISTRY }}
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Extract Docker metadata
+        id: meta
+        uses: docker/metadata-action@98669ae865ea3cffbcbaa878cf57c20bbf1c6c38
+        with:
+          images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
+          tags: |
+              type=ref,event=branch
+              type=ref,event=pr
+              type=semver,pattern={{version}}
+
+      - name: Build and push Docker image
+        uses: docker/build-push-action@ad44023a93711e3deb337508980b4b5e9bcdc5dc
+        with:
+          context: .
+          # Don't push on pull requests
+          push: ${{ github.event_name != 'pull_request' }}
```

Comment:
Unless this is the Dependabot issue do we need to prevent pushing from PRs?

Reply:
I was assuming that most developers would not want that because it would slow down their CI feedback circle. Depending on your CI pipeline, it could be pretty wasteful in terms of time and traffic to push every CI build to a registry so I assumed it was a sensible default.

In which scenarios would we want PRs to be pushed as well?

Reply:
I don't think the push has a high cost in terms of time, if we've already built the container as part of our PR we might as well push it and any layers will be reused in a future build after merging.

Reply:
For GHCR it would be fine to push an image but would need to be done through `pull_request_target` event to be able to use the `GITHUB_TOKEN` secret right?

Reply:
Yes, though the odd thing about the `pull_request_target` is that it pulls the target branch so you're not actually building the PR branch. This makes the artifacts created somewhat useless.  Ex: If your PR branch `foo` is targeting `main` `pull_request_target` will do a build against `main` instead of `foo`; which isn't what we want. However if we use the `pull_request` and the user who created the PR doesn't have write access the push will fail because the `GITHUB_TOKEN` is read-only.

Reply:
Ok so the condition on pull_request to push is correct here.

Reply:
I've removed that condition for now.